network/server.go

```diff
@@ -237,70 +179,291 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
+func setupLibp2pKey(secretsManager secrets.SecretsManager) (crypto.PrivKey, error) {
+	var key crypto.PrivKey
+
+	if secretsManager.HasSecret(secrets.NetworkKey) {
+		// The key is present in the secrets manager, read it
+		networkingKey, readErr := ReadLibp2pKey(secretsManager)
+		if readErr != nil {
+			return nil, fmt.Errorf("unable to read networking private key from Secrets Manager, %w", readErr)
 		}
 
-		// Check if at least one bootnode is specified
-		if len(s.config.Chain.Bootnodes) < MinimumBootNodes {
-			return ErrMinBootnodes
+		key = networkingKey
+	} else {
+		// The key is not present in the secrets manager, generate it
+		libp2pKey, libp2pKeyEncoded, keyErr := GenerateAndEncodeLibp2pKey()
+		if keyErr != nil {
+			return nil, fmt.Errorf("unable to generate networking private key for Secrets Manager, %w", keyErr)
 		}
 
-		// start discovery
-		s.discovery = &discovery{srv: s}
+		// Write the networking private key to disk
+		if setErr := secretsManager.SetSecret(secrets.NetworkKey, libp2pKeyEncoded); setErr != nil {
+			return nil, fmt.Errorf("unable to store networking private key to Secrets Manager, %w", setErr)
+		}
 
-		// try to decode the bootnodes
-		bootnodes := []*peer.AddrInfo{}
+		key = libp2pKey
+	}
 
-		for _, raw := range s.config.Chain.Bootnodes {
-			node, err := StringToAddrInfo(raw)
-			if err != nil {
-				return fmt.Errorf("failed to parse bootnode %s: %w", raw, err)
-			}
+	return key, nil
+}
 
-			if node.ID == s.host.ID() {
-				s.logger.Info("Omitting bootnode with same ID as host", "id", node.ID)
+// Start starts the networking services
+func (s *Server) Start() error {
+	s.logger.Info("LibP2P server running", "addr", common.AddrInfoToString(s.AddrInfo()))
 
-				continue
-			}
+	if setupErr := s.setupIdentity(); setupErr != nil {
+		return fmt.Errorf("unable to setup identity, %w", setupErr)
+	}
 
-			bootnodes = append(bootnodes, node)
+	// Set up the peer discovery mechanism if needed
+	if !s.config.NoDiscover {
+		// Parse the bootnode data
+		if setupErr := s.setupBootnodes(); setupErr != nil {
+			return fmt.Errorf("unable to parse bootnode data, %w", setupErr)
 		}
 
-		if setupErr := s.discovery.setup(bootnodes); setupErr != nil {
+		// Setup and start the discovery service
+		if setupErr := s.setupDiscovery(); setupErr != nil {
 			return fmt.Errorf("unable to setup discovery, %w", setupErr)
 		}
 	}
 
-	go func() {
-		if err := s.runJoinWatcher(); err != nil {
-			s.logger.Error(fmt.Sprintf("Unable to start join watcher service, %v", err))
-		}
-	}()
+	go s.runDial()
+	go s.checkPeerConnections()
 
 	// watch for disconnected peers
 	s.host.Network().Notify(&network.NotifyBundle{
 		DisconnectedF: func(net network.Network, conn network.Conn) {
 			go func() {
-				s.delPeer(conn.RemotePeer())
+				s.removePeer(conn.RemotePeer())
 			}()
 		},
 	})
 
 	return nil
 }
 
+// setupDiscovery Sets up the discovery service for the node
+func (s *Server) setupDiscovery() error {
```

Comment:
Why not move this process into discovery package?

Reply:
We can't exactly move it solely to the discovery package because it is dependent on multiple methods in the networking server, namely dial queue manipulation. To avoid having the discovery service set itself up, and to make sure the base networking server is in control of the setup, it was done in the networking package. The services just need to run.

I've moved out the init logic for both the identity and discovery service to a separate file so the `server.go` is much cleaner:

[1cd1083](https://github.com/0xPolygon/polygon-edge/pull/343/commits/1cd1083ae596b178b32cd09a2e39ae90778186ac)